data/src/main/java/org/apache/iceberg/data/orc/GenericOrcWriters.java

```diff
@@ -0,0 +1,612 @@
+import org.apache.orc.storage.ql.exec.vector.TimestampColumnVector;
+
+public class GenericOrcWriters {
+  private static final OffsetDateTime EPOCH = Instant.ofEpochSecond(0).atOffset(ZoneOffset.UTC);
+  private static final LocalDate EPOCH_DAY = EPOCH.toLocalDate();
+
+  private GenericOrcWriters() {
+  }
+
+  /**
+   * The interface for the conversion from Spark's SpecializedGetters to
+   * ORC's ColumnVectors.
+   */
+  interface Converter<T> {
+
+    Class<T> getJavaClass();
+
+    /**
+     * Take a value from the Spark data value and add it to the ORC output.
+     *
+     * @param rowId  the row in the ColumnVector
+     * @param data   either an InternalRow or ArrayData
+     * @param output the ColumnVector to put the value into
+     */
+    void addValue(int rowId, T data, ColumnVector output);
+  }
+
+  public static Converter<Boolean> booleans() {
+    return BooleanConverter.INSTANCE;
+  }
+
+  public static Converter<Byte> bytes() {
+    return ByteConverter.INSTANCE;
+  }
+
+  public static Converter<Short> shorts() {
+    return ShortConverter.INSTANCE;
+  }
+
+  public static Converter<Integer> ints() {
+    return IntConverter.INSTANCE;
+  }
+
+  public static Converter<LocalTime> times() {
+    return TimeConverter.INSTANCE;
+  }
+
+  public static Converter<Long> longs() {
+    return LongConverter.INSTANCE;
+  }
+
+  public static Converter<Float> floats() {
+    return FloatConverter.INSTANCE;
+  }
+
+  public static Converter<Double> doubles() {
+    return DoubleConverter.INSTANCE;
+  }
+
+  public static Converter<String> strings() {
+    return StringConverter.INSTANCE;
+  }
+
+  public static Converter<ByteBuffer> binary() {
+    return BytesConverter.INSTANCE;
+  }
+
+  public static Converter<UUID> uuids() {
+    return UUIDConverter.INSTANCE;
+  }
+
+  public static Converter<byte[]> fixed() {
+    return FixedConverter.INSTANCE;
+  }
+
+  public static Converter<LocalDate> dates() {
+    return DateConverter.INSTANCE;
+  }
+
+  public static Converter<OffsetDateTime> timestampTz() {
+    return TimestampTzConverter.INSTANCE;
+  }
+
+  public static Converter<LocalDateTime> timestamp() {
+    return TimestampConverter.INSTANCE;
+  }
+
+  public static Converter<BigDecimal> decimal18(TypeDescription schema) {
+    return new Decimal18Converter(schema);
+  }
+
+  public static Converter<BigDecimal> decimal38(TypeDescription schema) {
+    return Decimal38Converter.INSTANCE;
+  }
+
+  private static class BooleanConverter implements Converter<Boolean> {
+    private static final Converter<Boolean> INSTANCE = new BooleanConverter();
+
+    @Override
+    public Class<Boolean> getJavaClass() {
+      return Boolean.class;
+    }
+
+    @Override
+    public void addValue(int rowId, Boolean data, ColumnVector output) {
+      if (data == null) {
+        output.noNulls = false;
+        output.isNull[rowId] = true;
+      } else {
+        output.isNull[rowId] = false;
```

Comment:
This code is repeated for every converter. Consider doing something similar to the read side, where we put this logic in this Converter interface itself so the the implementations only need to think about non-null cases.
https://github.com/apache/iceberg/blob/809697a6bf56698b1bc52c5f039f2a190a58fed5/orc/src/main/java/org/apache/iceberg/orc/OrcValueReader.java#L25-L36